libafl_libfuzzer/libafl_libfuzzer_runtime/src/corpus.rs

```diff
@@ -0,0 +1,322 @@
+        match loaded_mapping.entry(idx) {
+            Entry::Occupied(mut e) => {
+                let &old = e.get();
+                let new = self.next_recency.fetch_add(1, Ordering::Relaxed);
+                e.insert(new);
+                loaded_entries.remove(&old);
+                loaded_entries.insert(new, idx);
+            }
+            Entry::Vacant(e) => {
+                // new entry! send it in
+                let new = self.next_recency.fetch_add(1, Ordering::Relaxed);
+                e.insert(new);
+                loaded_entries.insert(new, idx);
+            }
+        }
+        if loaded_entries.len() > self.max_len {
+            let idx = loaded_entries.pop_first().unwrap().1; // cannot panic
+            let cell = self.mapping.get(idx).ok_or_else(|| {
+                Error::key_not_found(format!("Tried to evict non-existent entry {idx}"))
+            })?;
+            let mut tc = cell.try_borrow_mut()?;
+            let _ = tc.input_mut().take();
+        }
+        Ok(())
+    }
+}
+
+impl<I> UsesInput for LibfuzzerCorpus<I>
+where
+    I: Input + Serialize + for<'de> Deserialize<'de>,
+{
+    type Input = I;
+}
+
+impl<I> Corpus for LibfuzzerCorpus<I>
+where
+    I: Input + Serialize + for<'de> Deserialize<'de>,
+{
+    fn count(&self) -> usize {
+        self.mapping.map.len()
+    }
+
+    fn add(&mut self, testcase: Testcase<Self::Input>) -> Result<CorpusId, Error> {
+        let idx = self.mapping.insert(RefCell::new(testcase));
+        let mut testcase = self.mapping.get(idx).unwrap().borrow_mut();
+
+        match testcase.file_path() {
+            Some(path) if path.canonicalize()?.starts_with(&self.corpus_dir) => {
+                // if it's already in the correct dir, we retain it
+            }
+            _ => {
+                let input = testcase.input().as_ref().ok_or_else(|| {
+                    Error::empty(
+                        "The testcase, when added to the corpus, must have an input present!",
+                    )
+                })?;
+                let name = input.generate_name(idx.into());
+                let path = self.corpus_dir.join(&name);
+
+                match input.to_file(&path) {
+                    Err(Error::File(e, _)) if e.kind() == ErrorKind::AlreadyExists => {
+                        // we do not care if the file already exists; in this case, we assume it is equal
+                    }
+                    res => res?,
+                }
+
+                // we DO NOT save metadata!
+
+                testcase.filename_mut().replace(name);
+                testcase.file_path_mut().replace(path);
+            }
+        };
+
+        self.touch(idx)?;
+        Ok(idx)
+    }
+
+    fn replace(
+        &mut self,
+        _idx: CorpusId,
+        _testcase: Testcase<Self::Input>,
+    ) -> Result<Testcase<Self::Input>, Error> {
+        unimplemented!("It is unsafe to use this corpus variant with replace!");
```

Comment:
Why not? Just create a new file and link to the new file, keep the old one around?

Reply:
This would mean that the old corpus entry would be loaded at start time. Better in this case (since this will only ever be used by libafl_libfuzzer_runtime) to keep it tight to requirements and fail fast if we do anything new :slightly_smiling_face:

Reply:
not sure about this one